File: testers/mariadb/java/src/java/testdb/MockRepoTest.java
package testdb;

import static org.junit.Assert.*;

import dev.typr.dsl.MockConnection;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.junit.Test;
import testdb.mariatest_identity.*;

/**
 * Tests for Mock repository implementations. Mock repos can be used for in-memory testing without a
 * database connection.
 */
public class MockRepoTest {

  @Test
  public void testMockInsert() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    // Insert without connection (mock doesn't need one)
    var row1 = mockRepo.insert(new MariatestIdentityRowUnsaved("Row 1"), null);
    var row2 = mockRepo.insert(new MariatestIdentityRowUnsaved("Row 2"), null);

    assertNotNull(row1.id());
    assertNotNull(row2.id());
    assertNotEquals(row1.id(), row2.id());
    assertEquals("Row 1", row1.name());
    assertEquals("Row 2", row2.name());
  }

  @Test
  public void testMockSelectById() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var inserted = mockRepo.insert(new MariatestIdentityRowUnsaved("Test Row"), null);

    var found = mockRepo.selectById(inserted.id(), null);
    assertTrue(found.isPresent());
    assertEquals(inserted.id(), found.get().id());
    assertEquals("Test Row", found.get().name());
  }

  @Test
  public void testMockSelectByIdNotFound() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var found = mockRepo.selectById(new MariatestIdentityId(999), null);
    assertFalse(found.isPresent());
  }

  @Test
  public void testMockSelectAll() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    mockRepo.insert(new MariatestIdentityRowUnsaved("Row A"), null);
    mockRepo.insert(new MariatestIdentityRowUnsaved("Row B"), null);
    mockRepo.insert(new MariatestIdentityRowUnsaved("Row C"), null);

    List<MariatestIdentityRow> all = mockRepo.selectAll(null);
    assertEquals(3, all.size());
  }

  @Test
  public void testMockUpdate() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var inserted = mockRepo.insert(new MariatestIdentityRowUnsaved("Original"), null);
    var updated = inserted.withName("Updated");

    boolean wasUpdated = mockRepo.update(updated, null);
    assertTrue(wasUpdated);

    var found = mockRepo.selectById(inserted.id(), null).orElseThrow();
    assertEquals("Updated", found.name());
  }

  @Test
  public void testMockUpdateNotFound() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var nonExistent = new MariatestIdentityRow(new MariatestIdentityId(999), "Test");
    boolean wasUpdated = mockRepo.update(nonExistent, null);
    assertFalse(wasUpdated);
  }

  @Test
  public void testMockDelete() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var inserted = mockRepo.insert(new MariatestIdentityRowUnsaved("To Delete"), null);

    boolean deleted = mockRepo.deleteById(inserted.id(), null);
    assertTrue(deleted);

    var found = mockRepo.selectById(inserted.id(), null);
    assertFalse(found.isPresent());
  }

  @Test
  public void testMockDeleteNotFound() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    boolean deleted = mockRepo.deleteById(new MariatestIdentityId(999), null);
    assertFalse(deleted);
  }

  @Test
  public void testMockUpsertInsert() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var newRow = new MariatestIdentityRow(new MariatestIdentityId(100), "New Row");
    var upserted = mockRepo.upsert(newRow, null);

    assertEquals(new MariatestIdentityId(100), upserted.id());
    assertEquals("New Row", upserted.name());

    // Verify it's in the mock storage
    var found = mockRepo.selectById(new MariatestIdentityId(100), null);
    assertTrue(found.isPresent());
  }

  @Test
  public void testMockUpsertUpdate() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var inserted = mockRepo.insert(new MariatestIdentityRowUnsaved("Original"), null);
    var toUpsert = inserted.withName("Upserted");
    var upserted = mockRepo.upsert(toUpsert, null);

    assertEquals(inserted.id(), upserted.id());
    assertEquals("Upserted", upserted.name());

    // Only one row should exist
    assertEquals(1, mockRepo.selectAll(null).size());
  }

  @Test
  public void testMockUpsertStreaming() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var row1 = new MariatestIdentityRow(new MariatestIdentityId(1), "Row 1");
    var row2 = new MariatestIdentityRow(new MariatestIdentityId(2), "Row 2");

    mockRepo.upsertBatch(List.of(row1, row2).iterator(), MockConnection.instance);

    var all =
        mockRepo.selectAll(null).stream()
            .sorted(Comparator.comparing(r -> r.id().value()))
            .toList();
    assertEquals(2, all.size());
    assertEquals("Row 1", all.get(0).name());
    assertEquals("Row 2", all.get(1).name());

    // Update via upsertStreaming
    var row1Updated = row1.withName("Row 1 Updated");
    var row2Updated = row2.withName("Row 2 Updated");

    mockRepo.upsertBatch(List.of(row1Updated, row2Updated).iterator(), MockConnection.instance);

    all =
        mockRepo.selectAll(null).stream()
            .sorted(Comparator.comparing(r -> r.id().value()))
            .toList();
    assertEquals(2, all.size());
    assertEquals("Row 1 Updated", all.get(0).name());
    assertEquals("Row 2 Updated", all.get(1).name());
  }

  @Test
  public void testMockUpsertBatch() {
    AtomicInteger idCounter = new AtomicInteger(1);
    var mockRepo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var row1 = new MariatestIdentityRow(new MariatestIdentityId(1), "Row 1");
    var row2 = new MariatestIdentityRow(new MariatestIdentityId(2), "Row 2");

    var result = mockRepo.upsertBatch(List.of(row1, row2).iterator(), null);
    assertEquals(2, result.size());

    // Verify they're stored
    assertEquals(2, mockRepo.selectAll(null).size());

    // Update via upsertBatch and verify returned rows
    var row1Updated = row1.withName("Row 1 Updated");
    var row2Updated = row2.withName("Row 2 Updated");

    var returned =
        mockRepo.upsertBatch(List.of(row1Updated, row2Updated).iterator(), null).stream()
            .sorted(Comparator.comparing(r -> r.id().value()))
            .toList();
    assertEquals(2, returned.size());
    assertEquals("Row 1 Updated", returned.get(0).name());
    assertEquals("Row 2 Updated", returned.get(1).name());
  }

  @Test
  public void testMockWithInterfacePolymorphism() {
    AtomicInteger idCounter = new AtomicInteger(1);

    // Use the interface type to demonstrate polymorphism
    MariatestIdentityRepo repo =
        new MariatestIdentityRepoMock(
            unsaved -> unsaved.toRow(() -> new MariatestIdentityId(idCounter.getAndIncrement())));

    var inserted = repo.insert(new MariatestIdentityRowUnsaved("Test"), null);
    var found = repo.selectById(inserted.id(), null);

    assertTrue(found.isPresent());
    assertEquals("Test", found.get().name());
  }
}

File: testers/mariadb/kotlin/src/kotlin/testdb/MockRepoTest.kt
package testdb

import dev.typr.dslkt.MockConnection
import org.junit.Assert.*
import org.junit.Test
import testdb.mariatest_identity.*
import java.util.concurrent.atomic.AtomicInteger

/**
 * Tests for mock repository implementations.
 */
class MockRepoTest {

    private fun createMockRepo(): Pair<MariatestIdentityRepoMock, AtomicInteger> {
        val idCounter = AtomicInteger(1)
        val mockRepo = MariatestIdentityRepoMock({ unsaved ->
            MariatestIdentityRow(
                id = MariatestIdentityId(idCounter.getAndIncrement()),
                name = unsaved.name
            )
        })
        return Pair(mockRepo, idCounter)
    }

    @Test
    fun testMockInsert() {
        val (mockRepo, _) = createMockRepo()

        val row1 = mockRepo.insert(MariatestIdentityRowUnsaved("Row 1"), MockConnection.instance)
        val row2 = mockRepo.insert(MariatestIdentityRowUnsaved("Row 2"), MockConnection.instance)

        assertNotNull(row1.id)
        assertNotNull(row2.id)
        assertNotEquals(row1.id, row2.id)
        assertEquals("Row 1", row1.name)
        assertEquals("Row 2", row2.name)
    }

    @Test
    fun testMockSelectById() {
        val (mockRepo, _) = createMockRepo()

        val inserted = mockRepo.insert(MariatestIdentityRowUnsaved("Test Row"), MockConnection.instance)

        val found = mockRepo.selectById(inserted.id, MockConnection.instance)
        assertNotNull(found)
        assertEquals(inserted.id, found!!.id)
        assertEquals("Test Row", found.name)
    }

    @Test
    fun testMockSelectByIdNotFound() {
        val (mockRepo, _) = createMockRepo()

        val found = mockRepo.selectById(MariatestIdentityId(999), MockConnection.instance)
        assertNull(found)
    }

    @Test
    fun testMockSelectAll() {
        val (mockRepo, _) = createMockRepo()

        mockRepo.insert(MariatestIdentityRowUnsaved("Row A"), MockConnection.instance)
        mockRepo.insert(MariatestIdentityRowUnsaved("Row B"), MockConnection.instance)
        mockRepo.insert(MariatestIdentityRowUnsaved("Row C"), MockConnection.instance)

        val all = mockRepo.selectAll(MockConnection.instance)
        assertEquals(3, all.size)
    }

    @Test
    fun testMockUpdate() {
        val (mockRepo, _) = createMockRepo()

        val inserted = mockRepo.insert(MariatestIdentityRowUnsaved("Original"), MockConnection.instance)
        val updated = inserted.copy(name = "Updated")

        val wasUpdated = mockRepo.update(updated, MockConnection.instance)
        assertTrue(wasUpdated)

        val found = mockRepo.selectById(inserted.id, MockConnection.instance)!!
        assertEquals("Updated", found.name)
    }

    @Test
    fun testMockUpdateNotFound() {
        val (mockRepo, _) = createMockRepo()

        val nonExistent = MariatestIdentityRow(MariatestIdentityId(999), "Test")
        val wasUpdated = mockRepo.update(nonExistent, MockConnection.instance)
        assertFalse(wasUpdated)
    }

    @Test
    fun testMockDelete() {
        val (mockRepo, _) = createMockRepo()

        val inserted = mockRepo.insert(MariatestIdentityRowUnsaved("To Delete"), MockConnection.instance)

        val deleted = mockRepo.deleteById(inserted.id, MockConnection.instance)
        assertTrue(deleted)

        val found = mockRepo.selectById(inserted.id, MockConnection.instance)
        assertNull(found)
    }

    @Test
    fun testMockDeleteNotFound() {
        val (mockRepo, _) = createMockRepo()

        val deleted = mockRepo.deleteById(MariatestIdentityId(999), MockConnection.instance)
        assertFalse(deleted)
    }

    @Test
    fun testMockUpsertInsert() {
        val (mockRepo, _) = createMockRepo()

        val newRow = MariatestIdentityRow(MariatestIdentityId(100), "New Row")
        val upserted = mockRepo.upsert(newRow, MockConnection.instance)

        assertEquals(MariatestIdentityId(100), upserted.id)
        assertEquals("New Row", upserted.name)

        val found = mockRepo.selectById(MariatestIdentityId(100), MockConnection.instance)
        assertNotNull(found)
    }

    @Test
    fun testMockUpsertUpdate() {
        val (mockRepo, _) = createMockRepo()

        val inserted = mockRepo.insert(MariatestIdentityRowUnsaved("Original"), MockConnection.instance)
        val toUpsert = inserted.copy(name = "Upserted")
        val upserted = mockRepo.upsert(toUpsert, MockConnection.instance)

        assertEquals(inserted.id, upserted.id)
        assertEquals("Upserted", upserted.name)
        assertEquals(1, mockRepo.selectAll(MockConnection.instance).size)
    }

    @Test
    fun testMockUpsertBatch() {
        val (mockRepo, _) = createMockRepo()

        val row1 = MariatestIdentityRow(MariatestIdentityId(1), "Row 1")
        val row2 = MariatestIdentityRow(MariatestIdentityId(2), "Row 2")

        mockRepo.upsertBatch(listOf(row1, row2).iterator(), MockConnection.instance)

        val all = mockRepo.selectAll(MockConnection.instance).sortedBy { it.id.value }
        assertEquals(2, all.size)
        assertEquals("Row 1", all[0].name)
        assertEquals("Row 2", all[1].name)

        val row1Updated = row1.copy(name = "Row 1 Updated")
        val row2Updated = row2.copy(name = "Row 2 Updated")

        mockRepo.upsertBatch(listOf(row1Updated, row2Updated).iterator(), MockConnection.instance)

        val allAfter = mockRepo.selectAll(MockConnection.instance).sortedBy { it.id.value }
        assertEquals(2, allAfter.size)
        assertEquals("Row 1 Updated", allAfter[0].name)
        assertEquals("Row 2 Updated", allAfter[1].name)
    }

    @Test
    fun testMockWithInterfacePolymorphism() {
        val idCounter = AtomicInteger(1)

        val repo: MariatestIdentityRepo = MariatestIdentityRepoMock({ unsaved ->
            MariatestIdentityRow(
                id = MariatestIdentityId(idCounter.getAndIncrement()),
                name = unsaved.name
            )
        })

        val inserted = repo.insert(MariatestIdentityRowUnsaved("Test"), MockConnection.instance)
        val found = repo.selectById(inserted.id, MockConnection.instance)

        assertNotNull(found)
        assertEquals("Test", found!!.name)
    }
}
